test/pytest/test_bridge.py

```diff
@@ -630,3 +635,282 @@ async def test_large_upload(event_loop, transport, tmp_path):
+
+        # with '/' appended, this should only open dirs
+        client = await FsInfoClient.open(transport, str(path) + '/')
+        assert await client.wait() == expected_state
+
+
+@pytest.mark.asyncio
+async def test_fsinfo_onlydir_watch(transport: MockTransport, fsinfo_test_cases: 'dict[Path, JsonObject]') -> None:
+    for path, expected_state in fsinfo_test_cases.items():
+        # note the order here: because our check for notdir is implemented
+        # inside of the bridge, we try inotify first and check for notdir
+        # second.  if systemd_ctypes supports this some day, it may change.
+        if path.name.startswith('no-r'):
+            # we need this one because we can't inotify files without +r
+            expected_state = fsinfo_err(errno.EACCES)
+        elif 'dir' not in path.name and 'error' not in expected_state:
+            # and this one to deal with not-dir
+            expected_state = fsinfo_err(errno.ENOTDIR)
+
+        # with '/' appended, this should only open dirs
+        client = await FsInfoClient.open(transport, str(path) + '/', watch=True)
+        assert await client.next_state() == expected_state
+        await client.close()
+
+
+@pytest.mark.asyncio
+async def test_fsinfo_watch_identity_changes(
+        transport: MockTransport, tmp_path: Path, fsinfo_test_cases: 'dict[Path, JsonObject]'
+) -> None:
+    # we will now point a symlink to the various possibilities to make sure the
+    # transitions are correctly handled
+    link = tmp_path / 'link'
+    client = await FsInfoClient.open(transport, link, watch=True)
+
+    # we didn't make a link yet, so...
+    assert (await client.next_state()) == fsinfo_err(errno.ENOENT)
+
+    # in watch mode we report errors a bit more sensitively: we also report
+    # them if we can't inotify the inode in question, which requires +r
+    # permission.  as such, we need to modify our 'no-r' tests:
+    for path in list(fsinfo_test_cases):
+        if path.name.startswith('no-r'):
+            fsinfo_test_cases[path] = fsinfo_err(errno.EACCES)
+        if path.name == 'dangling':
+            # this breaks our logic below because we transition from ENOENT to ENOENT
+            del fsinfo_test_cases[path]
+
+    possibilities = tuple(fsinfo_test_cases)
+    for from_file in possibilities:
+        for to_file in possibilities:
+            if from_file is to_file:
+                continue
+
+            # link to the original file and check state
+            link.symlink_to(from_file)
+            assert await client.next_state() == fsinfo_test_cases[from_file]
+
+            # generate some events immediately before we switch the link to
+            # ensure the event is suppressed
+            with contextlib.suppress(OSError):
+                from_file.touch()
+            with contextlib.suppress(OSError):
+                (from_file / 'a.txt').touch()
+            with contextlib.suppress(OSError):
+                (from_file / 'a.txt').unlink()
```

Comment:
Why the suppresses here? That feels odd in a test -- first of all, they shouldn't fail (it's our tmpdir), and second it would just fail later on anyway as the test setup isn't what the assertions below expect?

Reply:
`from_file` might not even be a directory... or we might not have write access to `touch()` it — we remove permissions from some of the files.

Reply:
Hmm, I'd prefer some explicit `.startswith('no-r')` or so to avoid surprises, but again: no strong opinion, feel free to just resolve this thread.

Reply:
i made the comment a bit clearer about why we expect the failures.